src/borg/cache.py

```diff
@@ -494,103 +673,104 @@ def create_master_idx(chunk_idx):
-                if self.progress:
+                        fetch_and_build_idx(archive_id, repository, self.cache.key, chunk_idx)
+                if self.cache.progress:
                     pi.finish()
             logger.info('Done.')
             return chunk_idx
 
         def legacy_cleanup():
             """bring old cache dirs into the desired state (cleanup and adapt)"""
             try:
-                os.unlink(os.path.join(self.path, 'chunks.archive'))
+                os.unlink(os.path.join(self.cache.path, 'chunks.archive'))
             except:
                 pass
             try:
-                os.unlink(os.path.join(self.path, 'chunks.archive.tmp'))
+                os.unlink(os.path.join(self.cache.path, 'chunks.archive.tmp'))
             except:
                 pass
             try:
                 os.mkdir(archive_path)
             except:
                 pass
 
-        self.begin_txn()
-        with cache_if_remote(self.repository) as repository:
+        self.cache.begin_txn()
+        with cache_if_remote(self.cache.repository) as repository:
             legacy_cleanup()
             # TEMPORARY HACK: to avoid archive index caching, create a FILE named ~/.cache/borg/REPOID/chunks.archive.d -
             # this is only recommended if you have a fast, low latency connection to your repo (e.g. if repo is local disk)
             self.do_cache = os.path.isdir(archive_path)
             self.chunks = create_master_idx(self.chunks)
 
+
+class AdhocChunksCache(ChunksCache):
+    def __init__(self, cache):
+        logger.warning('Note: --avoid-cache-sync is an experimental feature.')
+        super().__init__(cache=cache)
+        self.init_chunks()
+
+    def init_chunks(self):
+        # Explicitly set the initial hash table capacity to avoid performance issues
+        # due to hash table "resonance".
+        # Since we're creating an archive, add 10 % from the start
+        t0 = perf_counter()
+        num_requests = 0
+        num_chunks = len(self.cache.repository)
+        capacity = int(num_chunks / ChunkIndex.MAX_LOAD_FACTOR * 1.1)
+        self.chunks = ChunkIndex(capacity)
+        marker = None
+        pi = ProgressIndicatorPercent(total=num_chunks, msg="Downloading chunk list... %3.0f%%", msgid='cache.download_chunks')
+        while True:
+            num_requests += 1
+            result = self.cache.repository.list(limit=LIST_SCAN_LIMIT, marker=marker)
+            if not result:
+                break
+            pi.show(len(result))
+            marker = result[-1]
+            # All chunks from the repository have a refcount of MAX_VALUE, which is sticky,
+            # therefore we can't/won't delete them. Chunks we added ourselves in this transaction
+            # (e.g. checkpoint archives) are tracked correctly.
+            init_entry = ChunkIndexEntry(refcount=ChunkIndex.MAX_VALUE, size=0, csize=ChunkIndex.MAX_VALUE)
```

Comment:
csize=0 is another safe choice and has a more compact representation (1 byte vs 5 bytes).